VPLAY-11143: AAMP mp4Demux Integration

Vinish100 opened the pull request:
Reason for change: Integrate mp4Demux formally into AAMP
Test Procedure: Test with useMp4Demux=true and validate all types of Linear, VOD content in all apps.
Risks: Medium

Vinish100 committed:
VPLAY-11143: AAMP mp4Demux Integration

Reason for change: Integrate mp4Demux formally into AAMP
Test Procedure: Test with useMp4Demux=true and validate
all types of Linear, VOD content in all apps.
Risks: Low

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
42e926de99deb7237cd337c4dc30599a
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "this->mPauseInjector" without holding lock "InterfacePlayerRDK.mSourceSetupMutex". Elsewhere, "InterfacePlayerRDK.mPauseInjector" is written to with "InterfacePlayerRDK.mSourceSetupMutex" held 2 out of 2 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[middleware/InterfacePlayerRDK.cpp:2961](undefined//blob/undefined/middleware/InterfacePlayerRDK.cpp#L2961)

rdkcmf-jenkins commented on aampgstplayer.cpp:
```diff
@@ -679,7 +678,7 @@ void AAMPGstPlayer::NotifyInjectorToResume()
  *  @brief Inject stream buffer to gstreamer pipeline
  */
-bool AAMPGstPlayer::SendHelper(AampMediaType mediaType, const void *ptr, size_t len, double fpts, double fdts, double fDuration, bool copy, double fragmentPTSoffset, bool initFragment, bool discontinuity)
+bool AAMPGstPlayer::SendHelper(AampMediaType mediaType, MediaSample sample, bool copy, bool initFragment, bool discontinuity)
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
fe9f2273e65defd2c4ed33de6c15c0c0
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Big parameter passed by value
Passing parameter sample of type "MediaSample" (size 144 bytes) by value, which exceeds the low threshold of 128 bytes.

_Low Impact, CWE-398_
_PASS_BY_VALUE_



</details>

rdkcmf-jenkins commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -3000,12 +3001,12 @@ void InterfacePlayerRDK::SetPlayerName(std::string name)
  *  @brief Inject stream buffer to gstreamer pipeline
  */
-bool InterfacePlayerRDK::SendHelper(int type, const void *ptr, size_t len, double fpts, double fdts, double fDuration, double fragmentPTSoffset, bool copy, bool initFragment, bool &discontinuity, bool &notifyFirstBufferProcessed, bool &sendNewSegmentEvent, bool &resetTrickUTC, bool &firstBufferPushed)
+bool InterfacePlayerRDK::SendHelper(int type, MediaSample sample, bool copy, bool initFragment, bool &discontinuity, bool &notifyFirstBufferProcessed, bool &sendNewSegmentEvent, bool &resetTrickUTC, bool &firstBufferPushed)
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
1e4158b833e9def314bcc578b5ec79f4
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Big parameter passed by value
Passing parameter sample of type "MediaSample" (size 144 bytes) by value, which exceeds the low threshold of 128 bytes.

_Low Impact, CWE-398_
_PASS_BY_VALUE_



</details>

rdkcmf-jenkins commented on aampgstplayer.cpp:
```diff
@@ -1312,3 +1326,63 @@ void AAMPGstPlayer::StopMonitorAvTimer()
+		gstCodecInfo.info.audio.channelCount = codecInfo.mInfo.audio.mChannelCount;
+		gstCodecInfo.info.audio.sampleRate = codecInfo.mInfo.audio.mSampleRate;
+	}
+	playerInstance->SetStreamCaps((GstMediaType)type, gstCodecInfo);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
14f114131bffb59d660ee574460d9a04
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Uninitialized scalar variable
Using uninitialized value "gstCodecInfo.info" when calling "SetStreamCaps".

_High Impact, CWE-457_
_UNINIT_



</details>

rdkcmf-jenkins commented on aampgstplayer.cpp:
```diff
@@ -734,7 +733,8 @@ bool AAMPGstPlayer::SendHelper(AampMediaType mediaType, const void *ptr, size_t
 	{
-		privateContext->mBufferControl[mediaType].notifyFragmentInject(this, mediaType, fpts, fdts, fDuration, discontinuity);
+		// TODO: Accessing values from a moved object
+		privateContext->mBufferControl[mediaType].notifyFragmentInject(this, mediaType, sample.pts, sample.dts, sample.duration, discontinuity);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
1b3393775421469d767f721212b90964
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Using a moved object
"sample" is used after it has been already moved.

_High Impact, CWE-457_
_USE_AFTER_MOVE_



</details>

rdkcmf-jenkins commented on AampDemuxDataTypes.h:
```diff
@@ -0,0 +1,110 @@
+	AampCodecInfo(StreamOutputFormat format) : mCodecFormat(format), mIsEncrypted(false), mCodecData()
+	{
+		std::memset(&mInfo, 0, sizeof(mInfo));
+	}
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
b5905bcd8b9059a0fc7f63bb918e88ef
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Uninitialized scalar field
Non-static class member "mType" is not initialized in this constructor nor in any functions that it calls.

_Medium Impact, CWE-457_
_UNINIT_CTOR_



</details>

rdkcmf-jenkins commented on AampDemuxDataTypes.h:
```diff
@@ -0,0 +1,110 @@
+	AampCodecInfo() : mCodecFormat(FORMAT_INVALID), mIsEncrypted(false), mCodecData()
+	{
+		std::memset(&mInfo, 0, sizeof(mInfo));
+	}
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
7b292fab643e462f3577ba758c1dc275
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Uninitialized scalar field
Non-static class member "mType" is not initialized in this constructor nor in any functions that it calls.

_Medium Impact, CWE-457_
_UNINIT_CTOR_



</details>

rdkcmf-jenkins commented on test/aampcli/AampcliPlaybackCommand.cpp:
```diff
@@ -1096,20 +1097,61 @@ void PlaybackCommand::parse( const char *path )
+						if (samples.empty())
+						{
+							AAMPCLI_PRINTF("No samples found in file '%s'\n", path );
+							auto codecInfo = mp4Demux->getCodecInfo();
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c0fbc777f96d123d86df85b5bdc7fb57
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Use of auto that causes a copy
Using the "auto" keyword without an "&" causes the copy of an object of type "AampCodecInfo".

_Low Impact, CWE-none_
_AUTO_CAUSES_COPY_

## How to fix
 Use "const auto &" instead of "auto".

</details>

Vinish100 committed:
VPLAY-11143: AAMP Mp4Demux integration

Reason for change: Added function headers, split
up Mp4Demux.hpp to a .h and .cpp. Code cleanup
Test Procedure: Play clear and encrypted content using mp4demux
Risks: Low

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

rdkcmf-jenkins commented on mp4demux/MP4Demux.cpp:
```diff
@@ -0,0 +1,1114 @@
+		parseAuxInfo();
+	}
+	uint32_t entry_count = ReadU32();
+	for (uint32_t i = 0; i < entry_count; i++)
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
51c25cacb013cc97915b00e7a5025480
PRESENT
-->

# Coverity Issue - Logically dead code
Execution cannot reach the expression "i" inside this statement: "i++;".

_Medium Impact, CWE-561_
_DEADCODE_

rdkcmf-jenkins replied:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
51c25cacb013cc97915b00e7a5025480
PRESENT
-->

# Coverity Issue - Logically dead code
Execution cannot reach the expression "i" inside this statement: "i++;".

_Medium Impact, CWE-561_
_DEADCODE_

Vinish100 committed:
VPLAY-11143: AAMP Mp4Demux Integration

Reason for change: Ensure coding guidelines are
strictly followed. Code cleanup. Enforce encrypted
caps are persisted when processing clear samples
in between
Test Procedure: As mentioned in the ticket
Risks: None

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

lashmintha committed:
VPLAY-11143: AAMP mp4Demux Integration

Reason for change: Fix L1s with Mp4Demux
Test Procedure: Test with useMp4Demux=true and validate
all types of Linear, VOD content in all apps.
Risks: Low

Signed-off-by: lashmintha <lashminthaj17@gmail.com>

rdkcmf-jenkins commented on aampgstplayer.cpp:
```diff
@@ -1312,3 +1311,11 @@ void AAMPGstPlayer::StopMonitorAvTimer()
+	CodecInfo gstCodecInfo;
+	gstCodecInfo.codecFormat = (GstStreamOutputFormat)codecInfo.mCodecFormat;
+	gstCodecInfo.codecData = codecInfo.mCodecData;
+	playerInstance->SetStreamCaps((GstMediaType)type, (GstStreamOutputFormat)format, gstCodecInfo);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c48928b2f44d4658821846b683108d8d
PRESENT
-->

# Coverity Issue - Uninitialized scalar variable
Using uninitialized value "gstCodecInfo.info" when calling "SetStreamCaps".

_High Impact, CWE-457_
_UNINIT_

rdkcmf-jenkins replied:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c48928b2f44d4658821846b683108d8d
PRESENT
-->

# Coverity Issue - Uninitialized scalar variable
Using uninitialized value "gstCodecInfo.info" when calling "SetStreamCaps".

_High Impact, CWE-457_
_UNINIT_

Vinish100 committed:
VPLAY-11143: AAMP mp4Demux Integration

Reason for change: Fix review comments. Introduced
error detection and propogation in Mp4Demuxer. Code cleanup
Test Procedure: As mentioned in the ticket
Risks: Low

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

pstroffolino commented on mp4demux/AampMp4Demuxer.h:
```diff
@@ -0,0 +1,132 @@
+	 * @param[in] frameRate - rate per second
+	 * @return void
+	 */
+	void setFrameRateForTM (int frameRate) override { }
```
should spell out "TrickMode" rather than using "TM" abbreviation

Vinish100 replied:
Will take it in a different ticket. Too many changes in this one as-is

pstroffolino commented on mp4demux/AampMp4Demuxer.h:
```diff
@@ -0,0 +1,132 @@
+	/**
+	 * @brief optionally specify new pts offset to apply for subsequently injected TS media segments
+	 */
+	void setPtsOffset( double ptsOffset ) override { };
```
assuming this for non-invasive pts restamping?

Vinish100 replied:
Yes. We need to implement in Phase 2

varshnie committed:
VPLAY-11143: AAMP mp4Demux Integration

Reason for change:pipeline reconfigure is avoided if stream format is unknown
Procedure: As mentioned in the ticket
Risks: Low

Signed-off-by: varshnie <varshniblue14@gmail.com>

varshnie committed:
VPLAY-11143: AAMP mp4Demux Integration

Reason for change:Removed isobmff format setting hack
Procedure: As mentioned in the ticket
Risks: Low

Signed-off-by: varshnie <varshniblue14@gmail.com>

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
e5ebf547043be97cda734849d893ea67
PRESENT
-->

# Coverity Issue - Resource leak in object
Allocating memory by calling "new Configs".

_High Impact, CWE-401_
_CTOR_DTOR_LEAK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[middleware/InterfacePlayerRDK.cpp:77](undefined//blob/undefined/middleware/InterfacePlayerRDK.cpp#L77)

lashmintha committed:
VPLAY-11829 L1 Test coverage for Mp4demux

Reason for change: Included possible L1 test for validating Mp4demux functionality,
                     1. Added AampMp4DemuxTests tests which validates SendSegment() in possible scenario's
                     2. Added Mp4BoxParsingtests which validates all the box parsing pssh, saiz, senc, trun, tfhd, moof, tenc, ftyp

Test Procedure: Test with useMp4Demux=true and validate all types of Linear, VOD content in all apps.

Risks: Low

Signed-off-by: lashmintha <lashminthaj17@gmail.com>

Vinish100 committed:
VPLAY-11143: Mp4Demux integration in AAMP

Reason for change: Fix compilation issues due to rebase.
Update L1 tests to validate the demuxer and box parsing
code with sample mp4 data in Mp4DemuxTestData.h.
Added new  middleware L1 and minor refactoring. Removed magic
numbers in mp4demux. Unify codec, drmMetadata and PSSH data
structures in middleware and AAMP.
Test Procedure: Confirm mp4Demux works and L1 passes
Risks: None

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

DomSyna commented on middleware/InterfacePlayerRDK.h:
```diff
@@ -147,6 +146,100 @@ enum class InterfaceCB
 	startNewSubtitleStream // Add more events here if needed
 };
 
+// Codec information structure
```
This looks like duplicate definition of AampCodecInfo. This adds both a maintenance burden, potential for inconsistency.

Would it be better to to put things like this in a common header file ?

DomSyna replied:
oh AAMPCodecInfo has more fields in the audio / video is this intentional ?

Vinish100 replied:
No, that will be cleaned up

Vinish100 replied:
Done

jfagunde commented on middleware/GstUtils.h:
```diff
@@ -66,6 +66,7 @@ enum GstStreamOutputFormat
         GST_FORMAT_ISO_BMFF,        /**< ISO Base Media File format */
         GST_FORMAT_AUDIO_ES_MP3,    /**< MP3 Audio Elementary Stream */
         GST_FORMAT_AUDIO_ES_AAC,    /**< AAC Audio Elementary Stream */
+        GST_FORMAT_AUDIO_ES_AAC_RAW,/**< AAC Raw Audio Elementary Stream */
```
Would it be possible to move adding this new stream output format to its own pull request? It seems fairly self-contained and it would reduce this PR.

Vinish100 replied:
Done. VPLAY-12257

DomSyna commented on aampgstplayer.cpp:
```diff
@@ -1312,3 +1329,76 @@ void AAMPGstPlayer::StopMonitorAvTimer()
+ */
+void AAMPGstPlayer::SetStreamCaps(AampMediaType type, AampCodecInfo &&codecInfo)
+{
+	CodecInfo gstCodecInfo;
```
If aamp used the midlewares CodecInfo throughout we wouldn't need methods like this as its basically doing a move from AAMPCodecInfo to CodecInfo

If we can't use the same struct we could use static_assert to validate the structs are identical at compile time. Then in here we can just

std::memcpy(&gstCodecInfo, &codecInfo, sizeof(CodecInfo));
// Move codecData last, after all other fields have been accessed
gstCodecInfo.codecData = std::move(codecInfo.mCodecData);
playerInstance->SetStreamCaps(static_cast<GstMediaType>(type), gstCodecInfo);

DomSyna replied:
Just noticed the codecInfo structs are actually different.

Vinish100 replied:
Done

jfagunde commented on AampDemuxDataTypes.h:
```diff
@@ -0,0 +1,257 @@
+ * @struct AampMediaSample
+ * @brief Media sample structure
+ */
+struct AampMediaSample
```
How about moving the introduction of this structure to its own pull request, separate from the mp4demux code, to split this pull request into more manageable size ones?

DomSyna commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -5248,3 +5248,197 @@ double InterfacePlayerRDK::FlushTrack(int mediaType, double pos, double audioDel
+	GstStructure *metadata = NULL;
+	if (drmMetadata.isEncrypted)
+	{
+		metadata = gst_structure_new(
```
It looks like there is a lot or repetition here, could you refactor to use a helper function ?

maybe something like
static bool AddBufferFieldToStructure(GstStructure *metadata, const char *fieldName, const void *data, gsize size)
{
GstBuffer *buffer = CreateGstBufferWithData(data, size);
if (!buffer) return false;
gst_structure_set(metadata, fieldName, GST_TYPE_BUFFER, buffer, NULL);
gst_buffer_unref(buffer);
return true;
}

Vinish100 replied:
Done

DomSyna commented on mp4demux/MP4Demux.cpp:
```diff
@@ -0,0 +1,1285 @@
+{
+	switch (fourCC)
+	{
+		case MultiChar_Constant("avcC"):
```
Maybe iterate over
constexpr FourCCMapping codecMappings[] = {
	{ MultiChar_Constant("avcC"), FORMAT_VIDEO_ES_H264 },
	{ MultiChar_Constant("hvcC"), FORMAT_VIDEO_ES_HEVC },
	{ MultiChar_Constant("esds"), FORMAT_AUDIO_ES_AAC_RAW },
	{ MultiChar_Constant("dec3"), FORMAT_AUDIO_ES_EC3 }
};
to find the format ?
better from a SOLID pov

Vinish100 replied:
Done

DomSyna commented on mp4demux/MP4Demux.cpp:
```diff
@@ -0,0 +1,1285 @@
+		AampMediaSample& newSample = samples.back();
+		uint32_t sampleLen = defaultSampleSize;
+		uint32_t sampleDuration = defaultSampleDuration;
+		if (flags & 0x0100)
```
These flags would be better as defines and not magic numbers. Maybe copilot can sort it out ?

DomSyna commented on mediaprocessor.h:
violates ISP as it forces implementations to provide many methods they don't need
Probably not part of this work, but the code would benefit from breaking the class up into functional block

DomSyna commented on mp4demux/MP4Demux.cpp:
```diff
@@ -0,0 +1,1285 @@
+ */
+std::vector<AampPsshData> Mp4Demux::GetProtectionEvents()
+{
+	return std::move(protectionData);
```
std::vector<AampPsshData> result;
result.swap(protectionData);
return result;

Don't be tempted to return std::move(result); as it will stop the compiler doing NRVO and move is the natural fallback if NVRO isn't possible

Vinish100 replied:
As these are member variables, compiler doesn't attempt NVRO as per copilot. So a explicit std::move is required. Using rvalue references for return value is not suggested as well..

Vinish100 committed:
VPLAY-11143: AAMP mp4Demux Integration
Reason for change: Address review comments and cleaned
up some of the blocks/variables which were not used.
Test Procedure: Test with useMp4Demux=true and validate
all types of Linear, VOD content in all apps.
Risks: Low

Signed-off-by: Vinish100 <vinish.balan@gmail.com>

rdkcmf-jenkins commented on mp4demux/MP4Demux.cpp:
```diff
@@ -0,0 +1,1308 @@
+	handledEncryptedSamples(false),
+	codecInfo(GST_FORMAT_INVALID), parseError(MP4_PARSE_OK)
+{
+}
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
d84e307710c6c207005d6dbb2db7a8ca
PRESENT
-->

# Coverity Issue - Uninitialized scalar field
Non-static class member "heightFixed" is not initialized in this constructor nor in any functions that it calls.

_Medium Impact, CWE-457_
_UNINIT_CTOR_

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ba2c4c25ecf345c8ae0529fdaaa9b66d
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "this->mPauseInjector" without holding lock "InterfacePlayerRDK.mSourceSetupMutex". Elsewhere, "InterfacePlayerRDK.mPauseInjector" is written to with "InterfacePlayerRDK.mSourceSetupMutex" held 2 out of 2 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[middleware/InterfacePlayerRDK.cpp:2961](undefined//blob/undefined/middleware/InterfacePlayerRDK.cpp#L2961)
